smart-contract-verifier/smart-contract-verifier/src/lookup_methods/find_methods.rs

```diff
@@ -0,0 +1,139 @@
+use super::{
+    disassemble::{disassemble_bytecode, DisassembledOpcode},
+    method::Method,
+};
+use bytes::Bytes;
+use ethers_core::abi::Abi;
+use ethers_solc::sourcemap::SourceMap;
+use std::{collections::BTreeMap, iter::repeat};
+
+pub struct LookupMethodsRequest {
+    pub bytecode: Bytes,
+    pub abi: Abi,
+    pub source_map: SourceMap,
+    pub file_ids: BTreeMap<u32, String>,
+}
+
+pub struct LookupMethodsResponse {
+    pub methods: BTreeMap<String, Method>,
+}
+
+pub fn find_methods(request: LookupMethodsRequest) -> LookupMethodsResponse {
+    let methods = parse_selectors(request.abi);
+    let opcodes = disassemble_bytecode(&request.bytecode);
+    let opcodes = opcodes.as_slice();
+
+    let methods = methods
+        .into_iter()
+        .filter_map(|(func_sig, selector)| {
+            let func_index = match find_src_map_index(&selector, opcodes) {
+                Some(i) => i,
+                None => {
+                    tracing::warn!(func_sig, "function not found");
+                    return None;
+                }
+            };
+
+            tracing::debug!(func_sig, func_index, "found function");
+            let method = match Method::from_source_map(
+                selector,
+                &request.source_map,
+                func_index,
+                &request.file_ids,
+            ) {
+                Ok(m) => m,
+                Err(err) => {
+                    tracing::warn!(func_sig, err = err.to_string(), "failed to parse method");
+                    return None;
+                }
+            };
+            Some((hex::encode(selector), method))
+        })
+        .collect::<BTreeMap<String, Method>>();
```

Comment:
Nit: Just to make sure, that such explicit types declaration is intentional and you know that just a `collect()` will work here as well
```suggestion
        }).collect();
```